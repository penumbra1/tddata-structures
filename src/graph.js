export function createNode(key) {
  const neighbors = [];

  return {
    key,
    neighbors,
    addNeighbor(node) {
      neighbors.push(node);
    },
    print() {
      const neighborList = neighbors.map(n => n.key).join(" ");
      const line = `${this.key}${neighborList && ` => ${neighborList}`}`;

      return line;
    }
  };
}

export function createGraph(directed = false) {
  const nodes = [];
  const edges = [];

  return {
    directed,
    nodes,
    edges,
    addNode(key) {
      if (key == null) return null;

      const newNode = createNode(key);
      nodes.push(newNode);

      return newNode;
    },
    getNode(key) {
      return nodes.find(node => node.key === key) || null;
    },
    addEdge(node1Key, node2Key) {
      const [node1, node2] = [this.getNode(node1Key), this.getNode(node2Key)];

      if (!node1 || !node2) return null;

      node1.addNeighbor(node2);
      if (!this.directed) node2.addNeighbor(node1);

      const newEdge = `${node1Key}-${node2Key}`;

      edges.push(newEdge);

      return newEdge;
    },
    print() {
      return nodes.map(node => node.print()).join("\n");
    }
  };
}
