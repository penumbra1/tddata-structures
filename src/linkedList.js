export function createNode(value) {
  return {
    value,
    next: null
  };
}

// Singly linked list
export function createLinkedList() {
  return {
    head: null,
    tail: null,
    length: 0,
    get isEmpty() {
      return this.length === 0;
    },
    get(index) {
      if (index < 0 || index > this.length - 1) return null;

      if (index === 0) return this.head;
      if (index === this.length - 1) return this.tail;

      let current = this.head;

      let i = 0;
      while (i !== index) {
        current = current.next;
        i++;
      }

      return current;
    },
    delete(index) {
      if (index < 0 || index > this.length - 1) return null;

      let currentIndex = 0;
      let [previous, current, following] = [null, this.head, this.head.next];

      while (currentIndex < index) {
        [previous, current, following] = [current, following, following.next];
        currentIndex++;
      }

      if (previous === null) {
        // Current head will be removed
        this.head = following;
      } else {
        previous.next = following;
      }

      if (following === null) {
        // Current tail will be removed
        this.tail = previous;
      }

      this.length--;

      return current;
    },
    push(value) {
      const node = createNode(value);

      if (this.length === 0) {
        this.head = node;
      } else {
        this.tail.next = node;
      }

      this.tail = node;
      this.length++;
    },
    pop() {
      if (this.length === 0) {
        return null;
      }

      const oldTail = this.tail;
      // newTail will refer to the node previous to oldTail
      let newTail = this.head;

      if (this.length === 1) {
        this.head = null;
        newTail = null;
      } else {
        while (newTail.next !== this.tail) {
          newTail = newTail.next;
        }

        newTail.next = null;
      }

      this.tail = newTail;
      this.length--;

      return oldTail;
    },
    shift(value) {
      const node = createNode(value);

      node.next = this.head;
      this.head = node;
      if (this.length === 0) {
        this.tail = node;
      }
      this.length++;
    },
    unshift() {
      if (this.length === 0) {
        return null;
      }

      if (this.length === 1) {
        this.tail = null;
      }

      const oldHead = this.head;
      this.head = this.head.next;
      this.length--;

      return oldHead;
    },
    insertAt(index, value) {
      const node = createNode(value);

      if (index < 0 || index > this.length) return null;

      let currentIndex = 0;
      let [insertAfter, insertBefore] = [null, this.head];

      if (index === this.length) {
        // No need to traverse the list - fast forward to the end:
        currentIndex = this.length;
        [insertAfter, insertBefore] = [this.tail, null];
        this.tail = node;
      }

      while (currentIndex < index) {
        [insertAfter, insertBefore] = [insertBefore, insertBefore.next];
        currentIndex++;
      }

      if (insertAfter === null) {
        this.head = node;
      } else {
        insertAfter.next = node;
      }

      node.next = insertBefore;

      this.length++;

      return node;
    },
    find(value) {
      let current = this.head;
      let index = 0;

      while (current && current.value !== value) {
        current = current.next;
        index++;
      }

      if (current === null) return null;

      return index;
    },
    print() {
      let current = this.head;
      const values = [];

      while (current) {
        values.push(current.value);
        current = current.next;
      }

      return values.join(" => ");
    }
  };
}
