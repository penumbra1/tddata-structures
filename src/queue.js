export function createQueue() {
  const queue = [];

  return {
    enqueue(item) {
      queue.unshift(item);
    },
    dequeue() {
      return queue.pop();
    },
    peek() {
      return queue[queue.length - 1];
    },
    get length() {
      return queue.length;
    },
    get isEmpty() {
      return queue.length === 0;
    }
  };
}

export function createPriorityQueue() {
  const lowPriority = createQueue();
  const highPriority = createQueue();
  return {
    enqueue(item, isHighPriority = false) {
      if (isHighPriority) highPriority.enqueue(item);
      else lowPriority.enqueue(item);
    },
    dequeue() {
      if (!highPriority.isEmpty) return highPriority.dequeue();
      return lowPriority.dequeue();
    },
    peek() {
      if (!highPriority.isEmpty) return highPriority.peek();
      return lowPriority.peek();
    },
    get length() {
      return highPriority.length + lowPriority.length;
    },
    get isEmpty() {
      return highPriority.isEmpty && lowPriority.isEmpty;
    }
  };
}
