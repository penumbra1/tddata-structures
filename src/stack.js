export default function createStack(max = 5) {
  const stack = [];

  return {
    max,
    push(item) {
      if (stack.length < max) stack.push(item);
      else throw new Error("Stack overflow!");
    },
    pop() {
      return stack.pop();
    },
    peek() {
      return stack[stack.length - 1];
    },
    get isEmpty() {
      return stack.length === 0;
    },
    get length() {
      return stack.length;
    }
  };
}
