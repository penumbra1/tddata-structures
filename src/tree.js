export function createNode(key) {
  const children = [];

  return {
    key,
    children,
    addChild(childKey) {
      const child = createNode(childKey);
      children.push(child);
      return child;
    }
  };
}

export function createTree(rootKey) {
  if (rootKey == null) throw new Error("Creating a tree requires a root key");

  const root = createNode(rootKey);

  return {
    root,
    traverse(node, action, depth) {
      action(node, depth);

      if (node.children.length) {
        node.children.forEach(child => {
          this.traverse(child, action, depth + 1);
        });
      }
    },
    print() {
      let result = "";
      const addLine = (node, depth) => {
        result += depth
          ? `${"|  ".repeat(depth)}|--${node.key}\n`
          : `|--${node.key}\n`;
      };

      this.traverse(this.root, addLine, 0);

      return result;
    }
  };
}
