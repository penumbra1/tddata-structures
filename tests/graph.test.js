import { createNode, createGraph } from "../src/graph";

// Stub out addNeighbor to compare objects with methods via .toEqual
const mockNode = value => ({
  ...createNode(value),
  addNeighbor: expect.any(Function),
  print: expect.any(Function)
});

describe("Node", () => {
  let n;

  beforeEach(() => {
    n = createNode(1);
  });

  test("createNode should return a new node", () => {
    expect(n).toEqual(mockNode(1));
  });

  test("addNeighbor should add a node to the array of neighbors", () => {
    const neighbor = createNode(2);

    n.addNeighbor(neighbor);

    expect(n.neighbors).toEqual([neighbor]);
  });

  test("print should return the correct string representation", () => {
    expect(n.print()).toBe("1");

    n.addNeighbor(createNode(1));
    expect(n.print()).toBe("1 => 1");

    n.addNeighbor(createNode(2));
    expect(n.print()).toBe("1 => 1 2");
  });
});

describe("Graph", () => {
  let g;

  beforeEach(() => {
    g = createGraph();
  });

  test("createGraph should create an empty undirected or directed graph", () => {
    expect(g.directed).toBe(false);
    expect(g.nodes).toEqual([]);
    expect(g.edges).toEqual([]);

    g = createGraph(true);

    expect(g.directed).toBe(true);
    expect(g.nodes).toEqual([]);
    expect(g.edges).toEqual([]);
  });

  test("addNode should do nothing and return null if the key is invalid", () => {
    const invalidOne = g.addNode();
    const invalidTwo = g.addNode(null);

    expect(invalidOne).toBe(null);
    expect(invalidTwo).toBe(null);
    expect(g.nodes).toEqual([]);
    expect(g.edges).toEqual([]);
  });

  test("addNode should add a node to the graph if the key is valid", () => {
    const one = g.addNode(1);

    expect(one).toEqual(mockNode(1));
    expect(g.nodes).toEqual([one]);
    expect(g.edges).toEqual([]);
  });

  test("addNode should do nothing and return null if the key is repeated", () => {
    const [one, anotherOne] = [g.addNode(1), g.addNode(1)];

    expect(anotherOne).toBe(null);
    expect(g.nodes).toEqual([one]);
    expect(g.edges).toEqual([]);
  });

  test("getNode should return null if the key is not found", () => {
    expect(g.getNode()).toBe(null);
    expect(g.getNode(1)).toBe(null);

    g.addNode(1);
    expect(g.getNode(2)).toBe(null);
  });

  test("getNode should return the correct node", () => {
    g.addNode(1);
    const two = g.addNode(2);

    expect(g.getNode(2)).toBe(two);
  });

  test("addEdge should do nothing and throw if either node was not found", () => {
    g.addNode(1);
    const addEdgeOne = () => g.addEdge(1);
    const addEdgeTwo = () => g.addEdge(2, 3);

    expect(addEdgeOne).toThrow("Node not found (key: undefined)");
    expect(addEdgeTwo).toThrow("Node not found (key: 2)");
    expect(g.edges).toEqual([]);
    expect(g.getNode(1).neighbors).toEqual([]);
  });

  test("addEdge should add edge and neighbors correctly", () => {
    let [one, two] = [g.addNode(1), g.addNode(2)];

    let edge = g.addEdge(1, 2);

    expect(edge).toBe("1-2");
    expect(one.neighbors).toEqual([two]);
    expect(two.neighbors).toEqual([one]);

    g = createGraph(true);
    [one, two] = [g.addNode(1), g.addNode(2)];
    edge = g.addEdge(2, 1);

    expect(edge).toBe("2-1");
    expect(two.neighbors).toEqual([one]);
  });

  test("print should return the correct string representation", () => {
    expect(g.print()).toBe("");

    g.addNode(1);
    expect(g.print()).toBe("1");

    g.addNode(2);
    g.addNode(3);
    g.addEdge(3, 1);
    expect(g.print()).toMatchSnapshot();

    g = createGraph(true);
    g.addNode(1);
    g.addNode(2);
    g.addEdge(1, 2);

    expect(g.print()).toMatchSnapshot();
  });

  test("traverseBreadth should perform a breadth-first traversal", () => {
    g.addNode(1);
    g.addNode(2);
    g.addNode(3);
    g.addNode(4);
    g.addEdge(1, 2);
    g.addEdge(1, 3); // 1 => 2 3
    g.addEdge(2, 4); // 2 => 4
    const visited = [];

    g.traverseBreadth(1, node => visited.push(node.key));

    expect(visited.join("")).toMatch(/^1(23|32)4$/);
  });

  test("traverseBreadth should throw an error is startingKey is not found", () => {
    const visited = [];
    const traversal = () =>
      g.traverseBreadth(1, node => visited.push(node.key));

    expect(traversal).toThrow();
    expect(visited).toEqual([]);

    g.addNode(2);

    expect(traversal).toThrow();
    expect(visited).toEqual([]);
  });

  test("traverseDepth should perform a depth-first traversal", () => {
    g.addNode(1);
    g.addNode(2);
    g.addNode(3);
    g.addNode(4);
    g.addEdge(1, 2);
    g.addEdge(1, 3); // 1 => 2 3
    g.addEdge(2, 4); // 2 => 4
    const visited = [];

    g.traverseDepth(1, node => visited.push(node.key));

    expect(visited.join("")).toMatch(/^1(24|42)3$/);
  });

  test("traverseDepth should throw an error is startingKey is not found", () => {
    const visited = [];
    const traversal = () => g.traverseDepth(1, node => visited.push(node.key));

    expect(traversal).toThrow();
    expect(visited).toEqual([]);

    g.addNode(2);

    expect(traversal).toThrow();
    expect(visited).toEqual([]);
  });
});
