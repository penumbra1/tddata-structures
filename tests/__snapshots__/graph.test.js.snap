// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph print should return the correct string representation 1`] = `
"1 => 3
2
3 => 1"
`;

exports[`Graph print should return the correct string representation 2`] = `
"1 => 2
2"
`;
